bigint/src/lib.rs

```diff
@@ -1349,6 +1356,47 @@ impl Integer for BigUint {
     }
 }
 
+fn high_bits_to_u64(v: &BigUint) -> u64 {
+    match v.data.len() {
+        0   => 0,
+        1   => v.data[0] as u64,
+        _   => {
+            let mut bits = v.bits();
+            let mut ret = 0u64;
+            let mut ret_bits = 0;
+
+            for d in v.data.iter().rev() {
+                let digit_bits = (bits + big_digit::BITS - 1) %
+                    big_digit::BITS + 1;
```

Comment:
Hm. Actually, if adding to bits can overflow, bits itself can overflow -
it's a size_t, but size of an array in bits isn't guaranteed to fit in a
size_t.

On Jul 14, 2016 9:10 PM, "Josh Stone" <notifications@github.com> wrote:

In bigint/src/lib.rs
<https://github.com/rust-num/num/pull/204#discussion_r70922274>:

> @@ -1349,6 +1356,47 @@ impl Integer for BigUint {
>      }
>  }
>
> +fn high_bits_to_u64(v: &BigUint) -> u64 {
> +    match v.data.len() {
> +        0   => 0,
> +        1   => v.data[0] as u64,
> +        _   => {
> +            let mut bits = v.bits();
> +            let mut ret = 0u64;
> +            let mut ret_bits = 0;
> +
> +            for d in v.data.iter().rev() {
> +                let digit_bits = (bits + big_digit::BITS - 1) %
> +                    big_digit::BITS + 1;

Strictly speaking, it's not correct. It just trades a should-be-impossible
subtraction error for a possible-but-unlikely addition overflow. And
neither way gives a reasonable result, because bits==0 is a corner case
where digit_bits should be 0, otherwise bits -= bits_want will break. It's
not even more readable, because you had to wrap the line to fit the extra
addition.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/rust-num/num/pull/204/files/a1772f2a0d3e8fb7384a89322c93816b017fcc57#r70922274>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AB5r7mEtEkpH9qmvqF0B6TYOYBTKvh5Fks5qVxYpgaJpZM4JIeee>
.